maskrcnn_benchmark/modeling/roi_heads/box_head/loss.py

```diff
@@ -105,16 +108,79 @@ def subsample(self, proposals, targets):
 
         # distributed sampled proposals, that were obtained on all feature maps
         # concatenated via the fg_bg_sampler, into individual feature map levels
+        self.n_proposals_per_img = []
         for img_idx, (pos_inds_img, neg_inds_img) in enumerate(
             zip(sampled_pos_inds, sampled_neg_inds)
         ):
             img_sampled_inds = torch.nonzero(pos_inds_img | neg_inds_img).squeeze(1)
             proposals_per_image = proposals[img_idx][img_sampled_inds]
+            self.n_proposals_per_img.append(len(proposals_per_image))
             proposals[img_idx] = proposals_per_image
 
         self._proposals = proposals
         return proposals
 
+    def mining(self, class_logits, box_regression):
+        """
+        Similiar role as sumsample(), but return the rois with top loss.
+
+        Arguments:
+            class_logits (list[Tensor])
+            box_regression (list[Tensor])
+
+        Returns:
+            proposals (list[BoxList])
+        """
+
+        class_logits = cat(class_logits, dim=0)
+        box_regression = cat(box_regression, dim=0)
+        device = class_logits.device
+
+        if not hasattr(self, "_proposals"):
+            raise RuntimeError("subsample needs to be called before")
+
+        proposals = self._proposals
+
+        labels = cat([proposal.get_field("labels") for proposal in proposals], dim=0)
+        regression_targets = cat(
+            [proposal.get_field("regression_targets") for proposal in proposals], dim=0
+        )
+
+        classification_loss = F.cross_entropy(class_logits, labels, reduction='none')
+
+        # get indices that correspond to the regression targets for
+        # the corresponding ground truth labels, to be used with
+        # advanced indexing
+        sampled_pos_inds_subset = torch.nonzero(labels > 0).squeeze(1)
+        labels_pos = labels[sampled_pos_inds_subset]
+        if self.cls_agnostic_bbox_reg:
+            map_inds = torch.tensor([4, 5, 6, 7], device=device)
+        else:
+            map_inds = 4 * labels_pos[:, None] + torch.tensor(
+                [0, 1, 2, 3], device=device)
+
+        box_loss = smooth_l1_loss(
+            box_regression[sampled_pos_inds_subset[:, None], map_inds],
+            regression_targets[sampled_pos_inds_subset],
+            reduction='none',
+            beta=1,
+        ).sum(dim=1, keepdim=True)
+        ohem_loss = classification_loss
+        ohem_loss[sampled_pos_inds_subset[:, None]] += box_loss
```

Comment:
This operation maybe is wrong, which should be modifed into
`ohem_loss[sampled_pos_inds_subset[:, None]] = ohem_loss[sampled_pos_inds_subset[:, None]] + box_loss`

Reply:
Hi, 
Thanks for reviewing. Would you please explain why this is wrong? I just do not understand the difference.

Reply:
I am not sure for this, I have tried your code above, but I find it fail to add the box_loss to the correspond classification loss at ohem.  Maybe you can debug whether it's the same case.
Actually I changed the code
`        ohem_loss = classification_loss.clone()
        ohem_loss[sampled_pos_inds_subset[:, None]] = ohem_loss[sampled_pos_inds_subset[:, None]] + box_loss
`

Reply:
I would the very least not use an in-place operation, as it might break backpropagation depending on the operation.

You could replace it with something like what @WeihongM mentioned

Reply:
Thanks for reviewing this. I have modified as you suggested.